src/util/triggers.ts

```diff
@@ -37,24 +40,74 @@ export function parseTriggers(triggers: string, aliases = DEFAULT_ALIASES): Trig
 }
 
-const noopFn = () => {};
-
-export function listenToTriggers(renderer: any, nativeElement: any, triggers: string, openFn, closeFn, toggleFn) {
+export function observeTriggers(renderer: any, nativeElement: any, triggers: string, isOpenedFn: () => boolean) {
   const parsedTriggers = parseTriggers(triggers);
-  const listeners = [];
 
   if (parsedTriggers.length === 1 && parsedTriggers[0].isManual()) {
-    return noopFn;
+    return EMPTY;
   }
+  return new Observable<boolean>(subscriber => {
+    const listeners = [];
+    const openFn = () => subscriber.next(true);
+    const closeFn = () => subscriber.next(false);
+    const toggleFn = () => subscriber.next(!isOpenedFn());
 
-  parsedTriggers.forEach((trigger: Trigger) => {
-    if (trigger.open === trigger.close) {
-      listeners.push(renderer.listen(nativeElement, trigger.open, toggleFn));
-    } else {
-      listeners.push(
-          renderer.listen(nativeElement, trigger.open, openFn), renderer.listen(nativeElement, trigger.close, closeFn));
-    }
+    parsedTriggers.forEach((trigger: Trigger) => {
+      if (trigger.open === trigger.close) {
+        listeners.push(renderer.listen(nativeElement, trigger.open, toggleFn));
+      } else {
+        listeners.push(
+            renderer.listen(nativeElement, trigger.open, openFn),
+            renderer.listen(nativeElement, trigger.close, closeFn));
+      }
+    });
+
+    return () => { listeners.forEach(unsubscribeFn => unsubscribeFn()); };
   });
+}
 
-  return () => { listeners.forEach(unsubscribeFn => unsubscribeFn()); };
+export function subscribeTriggers(observable: Observable<boolean>, openFn, closeFn) {
+  const subscription = observable.subscribe(next => (next ? openFn() : closeFn()));
+  return () => subscription.unsubscribe();
+}
+
+export function triggerDelay(openDelay: number, closeDelay: number, isOpenedFn: () => boolean) {
+  return (input$: Observable<boolean>) => {
+    let pending = null;
+    const filteredInput$ = input$.pipe(
+        map(open => ({open})), filter(event => {
+          const currentlyOpen = isOpenedFn();
+          if (currentlyOpen !== event.open && (!pending || pending.open === currentlyOpen)) {
+            pending = event;
+            return true;
+          }
+          if (pending && pending.open !== event.open) {
+            pending = null;
+          }
+          return false;
+        }),
+        share());
+    const delayedOpen$ = filteredInput$.pipe(filter(event => event.open), delay(openDelay));
+    const delayedClose$ = filteredInput$.pipe(filter(event => !event.open), delay(closeDelay));
+    return merge(delayedOpen$, delayedClose$)
+        .pipe(
+            filter(event => {
+              if (event === pending) {
+                pending = null;
+                return event.open !== isOpenedFn();
+              }
+              return false;
+            }),
+            map(event => event.open));
+  };
+}
+
+export function listenToTriggers(
```

Comment:
Just as an alternative, while I was trying to read the code with a fresh look.

I think I'd prefer it to be like this:

```ts
  const parsedTriggers = parseTriggers(triggers);

  if (parsedTriggers.length === 1 && parsedTriggers[0].isManual()) {
    return () => {};
  }

  const subscription = observeTriggers(renderer, nativeElement, parsedTriggers, isOpenedFn)
    .pipe(triggerDelay(openDelay, closeDelay, isOpenedFn))
    .subscribe(open => (open ? openFn() : closeFn()));

  return () => subscription.unsubscribe();
```

Reasoning being:
- I see the whole flow undisturbed without jumping: parse → create observable → pipe through the delay → subscribe → call open/close
- `observeTriggers` **only** sets up an observable
- one function less → no need in `subscribeTriggers`